core/src/cgcloud/core/box.py

```diff
@@ -365,15 +370,86 @@ def prepare( self, ec2_keypair_globs, instance_type=None, image_ref=None,
             raise UserError( "The first key pair name can't be a glob." )
         self.ec2_keypairs = ec2_keypairs
         self.ec2_keypair_globs = ec2_keypair_globs
-
         kwargs = dict( instance_type=instance_type,
                        key_name=ec2_keypairs[ 0 ].name,
                        placement=self.ctx.availability_zone,
                        security_groups=security_groups,
-                       instance_profile_arn=self._get_instance_profile_arn( ) )
+                       instance_profile_arn=self._get_instance_profile_arn( ))
+
         self._populate_instance_creation_args( image, kwargs )
+        if options['price'] != None:
+            assert spot_support is True
+            # this means we are using a spot bid. We need a couple additional kwargs
+            self._optimize_bid(instance_type, options, kwargs)
+
+            kwargs['price']=options['price']
+
+            kwargs['launch_group']=self.ctx.namespace.replace("/","")+"_launch_group"
+            try: # There is no min/max count parameter for request_spot_instances, just count
+                kwargs['count']=kwargs['max_count']
+                del kwargs['max_count']
+                del kwargs['min_count']
+            except KeyError:
+                pass
         self.instance_creation_args = kwargs
 
+    def _choose_zone(self, bid, spot_data):
+        # returns the zone to put your spot request based on:
+        #    1) zones with prices currently under the bid
+        #    2) zones with the most stable price
+
+        ZoneTuple = namedtuple("ZoneTuple", ["name","price_deviation"])
+        def _sort_zones(bid, spot_data, zones):
+            # returns two lists of tuples of form: [ (zone.name, std_deviation)...]
+            # one for zones over the bid price and one for zones under bid price. Each are sorted by
+            # increasing standard deviation values
+            markets_under_bid, markets_over_bid = [], []
+            for zone in zones:
+                zlist = filter(lambda spot_history: spot_history.availability_zone == zone.name, spot_data)
+                std_deviation = numpy.std([spot_history.price for spot_history in zlist])
```

Comment:
I know its bad but I'd add a utility method to cgcloud-lib that does this, instead of using numpy. Just because it's such a big dependency.